### PR TITLE
Tech Notes

### LINKED ISSUE #147: Technical guidance
For Release 5.1, information in the Friendly Technical Notes for Providers will be moved to the relevant parts of the Code of Practice.

Affects
- Section 6 Logging Usage
- Section 7 Processing Rules for Underlying COUNTER Reporting Data
- Addition of information about restating data into a new Section 7.11
- Section 8 SUSHI for Automated Report Harvesting
- Appendix D: Guidelines for Implementation

### PR DESCRIPTION
**Pull request by tashalouizamc**
Pulled info from tech notes for providers into relevant sections of the COP

https://www.projectcounter.org/wp-content/uploads/2021/09/Release5.0.2_FG_Tech_v3.pdf

### CODE REVIEW

**Comment by tashalouizamc** [comment]
QUERY for @berndoberknapp : I think we can delete appendix d and recode the subsequent appendices. Do you agree?

**Review by tashalouizamc — COMMENTED** [review]
Thanks Bernd

File: source/05-delivery/index.rst [hunk 1]
```diff
@@ -7,32 +7,40 @@
...
+* Usage MUST be processed for the entire month before any usage for that month can be included in reports. If usage for a given month is not available yet, the report provider MUST NOT return usage for that month and MUST include exception 3031 in the report/response to indicate that usage is not ready for requested dates.
+* A minimum of the current year plus the prior 24 months of usage data MUST be available, unless the report provider is newly COUNTER compliant.
+* When report providers become compliant with a new release of the Code of Practice, they begin compiling usage compliant with the new release from the time they become compliant; and they MUST continue to provide the older usage that complies with the previous release(s) of the Code of Practice to fulfil the requirement.
+* The reports MUST allow the report consumer the flexibility to specify a date range, in terms of months, within the most recent 24-month period.
+* Report providers should provide reports on a per-customer basis. That is, if a business school has a separate customer ID from its parent university, the school and the university should be sent separate COUNTER reports.
```
Comment (berndoberknapp) [inline source/05-delivery/index.rst]: That is very problematic. In my experience the result will be that parts of the university usage also end up in the business school report, and some of the usage only ends up in the business school report. As a consequence it will be difficult or even impossible to determine the total usage for all subscriptions within the university. What actually would be required in such a situation would be a parent customer ID that allows to fetch a consolidated/deduplicated report, i.e. something like a consortium setup.

File: source/05-delivery/index.rst [hunk 2]
```diff
@@ -7,32 +7,40 @@
...
 
-  * In tabular form as either an Excel or a tab-separated-value (TSV) file, or both. Additional file formats that can be easily imported into spreadsheet programs without loss or corruption may be offered at the vendor's discretion.
-  * JSON formatted in accordance with the COUNTER_SUSHI API Specification (see :numref:`sushi` below).
 
-* The reports in JSON, TSV and other text formats MUST be encoded using UTF-8. Tabular reports in text formats SHOULD include a byte order mark, so that spreadsheet programs can automatically detect the encoding. JSON reports and other SUSHI server responses MUST NOT include a byte order mark (according to `RFC 8259, Section 8.1 <https://datatracker.ietf.org/doc/html/rfc8259#section-8.1>`_).
+Delivering JSON Reports
+"""""""""""""""""""""""
+
+* JSON formatted in accordance with the COUNTER_SUSHI API Specification (see :numref:`sushi` below).
```
Comment (berndoberknapp) [inline source/05-delivery/index.rst]: Add "The reports MUST be ..."?

File: source/05-delivery/index.rst [hunk 3]
```diff
@@ -7,32 +7,40 @@
...
+* The reports MUST be encoded using UTF-8. JSON reports and other SUSHI server responses MUST NOT include a byte order mark (according to `RFC 8259, Section 8.1 <https://datatracker.ietf.org/doc/html/rfc8259#section-8.1>`_).
 * Each report MUST be delivered as a separate file to facilitate automated processing of usage reports into ERM and usage consolidation systems. For clarity, multiple reports MUST NOT be included in the same Excel file as separate worksheets.
```
Comment (berndoberknapp) [inline source/05-delivery/index.rst]: This should be in the "Delivering Tabular Reports" section.

File: source/05-delivery/index.rst [hunk 4]
```diff
@@ -7,32 +7,40 @@
...
+* Reports MUST be available for harvesting via the COUNTER_SUSHI API within 4 weeks of the end of the reporting period.
+
+
+Delivering Tabular Reports
+""""""""""""""""""""""""""
+
+* The reports MUST be encoded using UTF-8. Tabular reports in text formats SHOULD include a byte order mark, so that spreadsheet programs can automatically detect the encoding. .
```
Comment (berndoberknapp) [inline source/05-delivery/index.rst]: Remove additional point.

File: source/05-delivery/index.rst [hunk 5]
```diff
@@ -7,32 +7,40 @@
...
+* In tabular form as either an Excel or a tab-separated-value (TSV) file, or both. Additional file formats that can be easily imported into spreadsheet programs without loss or corruption may be offered at the vendor's discretion.
```
Comment (berndoberknapp) [inline source/05-delivery/index.rst]: Add "The reports MUST be delivered ..."?

File: source/05-delivery/index.rst [hunk 6]
```diff
@@ -7,32 +7,40 @@
...
+  * The report interface MUST offer all Standard Views of the COUNTER Reports the report provider is required to provide, and Standard Views options MUST automatically apply the REQUIRED filter and configuration options and not allow the report consumer to alter the filters or configuration options except for the usage begin and end dates.
   * The date range fields on the user interface MUST default to the latest month with complete usage. For example, if the current date is 15 May 2024 and April usage has been processed, the begin date would default to 01 April 2024 and the end date would default to 30 April 2024. If the April usage has not yet been processed, the start and end dates would default to 01 March 2024 and 31 March 2024.
   * COUNTER Reports must include the option to Exclude_Monthly_Details. Item Reports must include the options to Include_Parent_Details and Include_Component_Details (see :numref:`filters-attributes` for details).
```
Comment (berndoberknapp) [inline source/05-delivery/index.rst]: Add "if reporting on components is supported" for Include_Component_Details?

File: source/06-logging/01-log-file-analysis.rst [hunk 7]
```diff
@@ -5,10 +5,16 @@
...
-* The data is on the organization’s own servers and is in a standard, rather than a proprietary, format. This makes it easy for an organization to switch programs later, use several different programs, and analyse historical data with a new program.
-* Log files contain information on visits from search engine spiders. Although these MUST NOT be reported as part of user activity, it is useful information for search engine optimization.
+Most web servers produce log files by default in a pre-defined format which may differ by server type. The log file data needed for a COUNTER implementation should be available without changes to the platform.
+
+As log file data is held on servers in a standard format, report providers can use a variety of analytics programmes and receive consistent results over time. Log files are also independent of users’ web browsers, meaning report providers can reliably track all activity for the purposes of COUNTER reporting.
+
+Report providers wishing to learn more about log files should read the Amazon Web Service documentation (https://docs.aws.amazon.com/).
```
Comment (berndoberknapp) [inline source/06-logging/01-log-file-analysis.rst:14]: I noticed that you usually put URLs in round brackets which differs from what we've done so far. This should be made consistent, probably in the clean up pull request.

File: source/06-logging/02-page-tagging.rst [hunk 8]
```diff
@@ -5,13 +5,13 @@
...
-* Page tagging can report on events that do not involve a request to the web server.
-* Page tagging is available to companies who do not have access to their own web servers.
-* The page-tagging service manages the process of assigning cookies to visitors; with log file analysis, the server must be configured to do this.
-* Recently page tagging has become a standard in web analytics.
-* Log file analysis is almost always performed in-house. Page tagging can be done in-house, but is more often provided as a third-party service. The cost differences between these two models can also be a consideration.
+Page tags are small pieces of code embedded in each page of a platform website. Data is gathered via these tags and passed to a database. This data can then be manipulated and stored, allowing complete control over how the data is represented.
+
+A key difference between log file analysis and page tagging is that with page tags a usage count is activated by opening the page in the browser, not by requesting it from the server. This means page tagging is likely to offer a more accurate reflection of usage, because cached pages are counted in the same way as server calls.
```
Comment (berndoberknapp) [inline source/06-logging/02-page-tagging.rst]: In my experience many browsers block at least some page tagging services , which can be problematic. When we switched from log file analysis to Matomo for our catalog, the result was a sharp decrease in the usage numbers.

File: source/07-processing/01-http-status-codes.rst [hunk 9]
```diff
@@ -5,4 +5,4 @@
 HTTP Status Codes
 -----------------
 
-Only successful and valid requests MUST be counted. For web server log files successful requests are those with specific HTTP status codes (200 and 304). The standards for HTTP status codes are defined and maintained by the `IETF HTTP working group <https://datatracker.ietf.org/wg/httpbis/documents/>`_ in a series of RFCs (most notably `RFC 7231 <https://tools.ietf.org/html/rfc7231.html>`_). If key events are used, their definition MUST match the HTTP standards. (For more information see `The Friendly Guide to Release 5: Technical Notes for Content Providers <https://www.projectcounter.org/appendix-d-guidelines-implementation/>`_.)
+Only successful and valid requests MUST be counted. For web server log files successful requests are those with specific HTTP status codes (200 and 304). The standards for HTTP status codes are defined and maintained by the `IETF HTTP working group <https://datatracker.ietf.org/wg/httpbis/documents/>`_ in a series of RFCs (most notably `RFC 7231 <https://tools.ietf.org/html/rfc7231.html>`_). If key events are used, their definition MUST match the HTTP standards.
```
Comment (berndoberknapp) [inline source/07-processing/01-http-status-codes.rst]: The reference should be updated, since RFC 7231 has been obsoleted by RFC 9110.

File: source/07-processing/04-counting-unique-titles.rst [hunk 10]
```diff
@@ -7,9 +7,9 @@
 Counting Unique Titles
 ----------------------
 
-Some COUNTER Metric_Types count the number of unique titles that had a certain activity, such as a Unique_Title_Requests or Unique_Title_Investigations.
+Some COUNTER Metric_Types count the number of unique book titles that had a certain activity, namely Unique_Title_Requests or Unique_Title_Investigations.
 
-For the purpose of COUNTER metrics, a title represents the parent work that the item is part of. When the item is a chapter or section, the title is the book. The title MUST be identified using a unique identifier (e.g. an ISBN for a book) regardless of format (e.g. PDF or HTML).
+For the purpose of COUNTER metrics, a title represents the book (parent work) that the item (book chapter or other segment) is part of. The title MUST be identified using a unique identifier (e.g. an ISBN for a book) regardless of format (e.g. PDF or HTML).
```
Comment (berndoberknapp) [inline source/07-processing/04-counting-unique-titles.rst:12]: What about reference works?

File: source/07-processing/06-searches.rst [hunk 11]
```diff
@@ -2,8 +2,22 @@
...
+
+A search is counted any time the platform executes a search to retrieve a new set of results. This means that platforms performing multiple searches (e.g. search for exact match, search for words in subject, general search) to return a single set of results ordered by relevancy must only count a single search, not multiple searches.
+
+Design options and activities that do count as separate searches include:
+
+* Bento-box or multi-tab user interfaces, where multiple searches are conducted to retrieve and present multiple result sets.
+* Refinement of a set of search results by faceting, where applying a facet or filter requires the search to be re-run.
+* Browsing through a topics list or subject authority file, where clicking on the topic or subject conducts a search to present a set of search results.
+
+Note that link resolution MUST NOT counts as a search.
```
Comment (berndoberknapp) [inline source/07-processing/06-searches.rst]: count (not counts)

File: source/07-processing/07-federated-searches.rst [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+Any searches generated from a federated search system MUST be included in the separate Searches_Federated counts within Database Reports and MUST NOT be included in the Searches_Regular or Searches_Automated counts.
+
+The most common ways to recognize federated and automated search activity are as follows:
+
+* A federated search engine may be using its own dedicated IP address, which can be identified and used to separate out the activity.
+* If the standard HTML interface is being used (e.g. for screen scraping), the browser ID within the web log files can be used to identify the activity as coming from a federated search.
+* For Z39.50 activity (http://www.niso.org/standards/resources/Z39.50_Resources), authentication is usually through a username/password combination. Create a unique username/password that just the federated search engine will use.
+* If an API (application programming interface) gateway is available, set up an instance of the gateway that is for the exclusive use of federated search tools. It is recommended you also require the federated search engine to include an identifying parameter when making requests to the gateway.
```
Comment (berndoberknapp) [inline source/07-processing/07-federated-searches.rst]: All searches via Z39.50 and APIs must be counted as Searches_Federated, since the results are not presented in the host UI! So there is no need to distinguish specific clients or setup a seperate API endpoint.

File: source/07-processing/07-federated-searches.rst [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+Where federated or automated usage is genuine user-driven usage, in the context of Text & Data Mining, the Access_Method ‘TDM’ attribute should be used. This allows users of the resultant reports to distinguish automated usage from more traditional (‘regular’) usage.
```
Comment (berndoberknapp) [inline source/07-processing/07-federated-searches.rst]: There should be no single (or double) quotes for enumerated values. Suggestions: Access_Method=TDM or Access_Method TDM value (in both cases not attribute), and simply Regular instead of ('regular').

File: source/07-processing/07-federated-searches.rst [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+COUNTER has lists of federated search tools in Appendix G (See :ref:`Appendix G <appendix-g>`), separate from a list of robots which is reviewed and updated on a regular basis and which can be found at: https://github.com/atmire/COUNTER-Robots.
```
Comment (berndoberknapp) [inline source/07-processing/07-federated-searches.rst:23]: I assume the URL will be fixed by a separate pull request.

File: source/07-processing/11-restating-data.rst [hunk 15]
```diff
@@ -0,0 +1,20 @@
+.. The COUNTER Code of Practice Release 5 © 2017-2023 by COUNTER
+   is licensed under CC BY-SA 4.0. To view a copy of this license,
+   visit https://creativecommons.org/licenses/by-sa/4.0/
+
+Restating data
```
Comment (berndoberknapp) [inline source/07-processing/11-restating-data.rst]: Data

File: source/07-processing/11-restating-data.rst [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+--------------
+
+
+Retrospective Reporting of Errors
+"""""""""""""""""""""""""""""""""
+
+If errors are identified in COUNTER reports, the report provider must correct the errors within three months of their discovery. The report provider must alert the Project Director so that a notification may be included in the COUNTER Registry. Report providers should also directly inform their customers of the corrections.
```
Comment (berndoberknapp) [inline source/07-processing/11-restating-data.rst]: MUST instead of both must? SHOULD instead of should?

COUNTER Project Director, maybe with email address? This should be consistent thoughout the CoP.

File: source/07-processing/11-restating-data.rst [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+
+
+Reporting Usage When Journal Titles Change
+""""""""""""""""""""""""""""""""""""""""""
+
+When the title of a journal is changed but the identification code (ISSN or DOI) stays the same, report providers should consider the journal to be a single ‘Title’ for the purposes of COUNTER reporting. COUNTER Reports and Standard Views of the COUNTER Reports should be provided against the new title, with the original title being dropped from the reports.
+
+If a new DOI or ISSN is allocated to the journal when the title changes, Report providers should consider the journal to be two ‘Titles’ and provide usage information separately. Report providers should not report usage for the same period under both the old and new DOI or ISSN. Any report generated for the old DOI or ISSN should show zero usage from the month in which the new DOI or ISSN takes effect.
```
Comment (berndoberknapp) [inline source/07-processing/11-restating-data.rst]: The single quotes also should be removed in this section.

SHOULD instead of should? The "should not" probably should be "MUST NOT", because not following that rule would result in double reporting.

File: source/07-processing/12-usage-spikes.rst [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+Abnormal Spikes in Usage
+------------------------
+
+What is regarded as an abnormal spike in usage can vary from one report provider and report consumer to another and there are many occasions on which exceptionally high usage in a month is genuine, so COUNTER does not have a strict protocol for dealing with usage spikes. The following approaches will provide an indication of possible abnormal usage or another unusual event. These should only be as a prompt for human intervention to take a closer look at the numbers, rather than any automated cut-off of access.
+
+
+Positive Spikes in Usage
+""""""""""""""""""""""""
+
+Reported usage may be considered as a positive spike if, in a specific month, the reported usage by a report consumer is at least one hundred units of measurement greater than 300% above the previous twelve-month average.
```
Comment (berndoberknapp) [inline source/07-processing/12-usage-spikes.rst]: I'm wondering about the "above". Would that be 3\*average+100 or 4\*average+100?

File: source/08-sushi/index.rst [hunk 19]
```diff
@@ -7,7 +7,7 @@
 SUSHI for Automated Report Harvesting
 =====================================
 
-Report providers MUST support automatic harvesting of COUNTER reports via the COUNTER_SUSHI API. The specification for the RESTful COUNTER_SUSHI API is maintained by COUNTER on SwaggerHub:
+The Standardized Usage Statistics Harvesting Initiative (SUSHI) protocol is designed to simplify the gathering of usage statistics by librarians, and report providers MUST support automatic harvesting of COUNTER reports via the COUNTER_SUSHI API. The specification for the RESTful COUNTER_SUSHI API is maintained by COUNTER on SwaggerHub:
```
Comment (berndoberknapp) [inline source/08-sushi/index.rst]: We've decided to not use "Standardized Usage Statistics Harvesting Initiative" for the R5 SUSHI, since when you look for information about this you most likely end up on the NISO page which is confusing, instead SUSHI for R5 is short for COUNTER_SUSHI API (see the SUSHI glossary entry).

report consumers instead of librarians?

**Review by berndoberknapp — CHANGES_REQUESTED** [review]
Please see the inline comments.

The "Access to Usage for Consortia" section should be section 5.3, not 5.2.1.

File: source/07-processing/04-counting-unique-titles.rst (hunk 10 above)
Comment (tashalouizamc) [reply source/07-processing/04-counting-unique-titles.rst:12]: I am going to have to pick this up in clean up because of the merge conflicts

File: source/07-processing/07-federated-searches.rst (hunk 14 above)
Comment (tashalouizamc) [reply source/07-processing/07-federated-searches.rst:23]: I've not migrated the robots list yet, if that's the URL you're referring to?

File: source/07-processing/07-federated-searches.rst [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+
+Any searches generated from a federated search system MUST be included in the separate Searches_Federated counts within Database Reports and MUST NOT be included in the Searches_Regular or Searches_Automated counts.
+
+The most common ways to recognize federated and automated search activity are as follows:
+
+* A federated search engine may be using its own dedicated IP address, which can be identified and used to separate out the activity.
+* If the standard HTML interface is being used (e.g. for screen scraping), the browser ID within the web log files can be used to identify the activity as coming from a federated search.
+* All searches via APIs and Z39.50 activity (http://www.niso.org/standards/resources/Z39.50_Resources) must be counted as Searches_Federated, as the results are not presented in the platform user interface.
+* If an API gateway is available, set up an instance of the gateway that is for the exclusive use of federated search tools. It is recommended you also require the federated search engine to include an identifying parameter when making requests to the gateway.
```
Comment (berndoberknapp) [inline source/07-processing/07-federated-searches.rst:19]: Was it intentional that you've removed the recommendation for Z39.50 but kept the one for the API gateway?

**Review by berndoberknapp — APPROVED** [review]
Note that I've fixed the section/file structure for section 5, fixed the first paragraph for Delivering Tabular Reports and changed some additional should/must to upper case. I'll merge the pull request now, I think further changes could be addressed in the clean up pull request.

File: source/07-processing/07-federated-searches.rst (hunk 14 above)
Comment (berndoberknapp) [reply source/07-processing/07-federated-searches.rst:23]: Yes.

**Comment by berndoberknapp** [comment]
> I think we can delete appendix d and recode the subsequent appendices. Do you agree?

Yes. This should be done in a separate pull request as part of the clean up when we've commited all the other pull requests to avoid conflicts.
